### PR TITLE
fix(storage): consistent case-insensitive email identity + surface absorbed re-logins (#171)

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

Closes the two real (non-blocking) edge cases the **deep release-readiness stress test** found in the #171 account-recovery layers, plus a **third same-class divergence** the architect review surfaced. All three are email-identity consistency issues; none broke healthy-pool recovery, but they're worth fixing before cutting 6.3.4.

## The three fixes

**1. Email case divergence (storage vs detectors).** `lib/storage/identity.ts` compared emails case-*sensitively* (trim only) in the dedup/merge layer, while `lib/accounts/stale-state.ts` `findDisabledTokenSourceDuplicates` lowercases. So `User@Example.com` (org) + `user@example.com` (disabled token re-login) **escaped storage dedup yet were flagged as removable** by codex-doctor/health — the two #171 layers disagreed on identity. Added `sameEmailIdentity` (trim+lowercase) used only at the email comparison site, and lowercased the dedup-by-email map key. **`accountId`/`organizationId` comparisons remain case-sensitive** (sameOptionalIdentity untouched) so #64 multi-org / #85 workspace preservation are not weakened.

**2. Surfacing blind spot.** A user-disabled org account that absorbs a newer *enabled* token re-login correctly stays disabled (fail-closed), but was then skipped by all three #171 diagnostics — the user got no signal that a fresh login landed on a disabled slot. New read-only `findDisabledAccountsWithFreshCredential` (disabled + non-empty access token + unexpired) surfaced by codex-doctor (finding `disabled-account-fresh-credential`) and codex-health (text + `disabledWithFreshCredentialSlots` JSON), pointing to re-enable in the accounts file. **Never re-enables anything.**

**3. Third divergence (review-found).** `login-runner.ts buildIdentityIndexes` keyed `byEmailNoOrg` with `trim()` only, while its own lookup (`sanitizeEmail`) and `getExactIdentityKey` lowercase — so a mixed-case re-login could miss its existing no-org entry and append a duplicate. Aligned to `sanitizeEmail`.

## Tests
- 5 unit tests for `findDisabledAccountsWithFreshCredential` (disabled+fresh / enabled / expired / token-less / ordering).
- `test/storage.test.ts`: `User@Example.com` + `user@example.com` collapse on save/load to one enabled org account.
- `test/login-runner.test.ts`: behavioral guarantee — a mixed-case re-login merges to 1 account; a control with distinct emails stays 2 (proves non-vacuous).
- **Mutation teeth-checks** confirmed for fixes 1 and 2 (reverting each fails the corresponding test).

## Verification
- Full suite **2549 passed**, 1 skipped. `tsc --noEmit`, `eslint`, `npm run build` clean. `npm audit`: 0 vulnerabilities.

## Release note
This is the last cleanup before cutting **6.3.4** via release-please (all commits since v6.3.3 are `fix:`/`test:` → patch bump). The `Error: Aborted` *root trigger* (#176) remains intentionally open — only its symptom was typed in #178.

Relates to #171.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Bug Fixes
- Fixed account deduplication that could leave the canonical account disabled after merging duplicates
- Improved email matching during deduplication and re-login handling to be case-insensitive
- Enhanced Codex Doctor and Codex Health diagnostics to surface disabled accounts with fresh login credentials and prompt re-enablement

## Tests
- Added coverage for the new deduplication behavior and the fresh-credential disabled-account diagnostics

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

this pr closes three email-identity consistency gaps found during stress testing of the #171 account-recovery layers: case-sensitive email comparison in storage dedup, a trim-only index key in `login-runner.ts`'s `buildIdentityIndexes`, and a silent blind spot where a fresh credential on a disabled slot produced no diagnostic.

- **email case normalization**: `sameEmailIdentity` (trim+lowercase) now used in `isLegacyRefreshTokenDuplicate`, and `deduplicateAccountsByEmail` keys its email map with `.toLowerCase()`, so `User@Example.com` and `user@example.com` now collapse correctly in both the storage merge and the email-only dedup paths.
- **login-runner index fix**: `byEmailNoOrg` keyed via `sanitizeEmail(account.email)` (trim+lowercase+`@`-validation) instead of bare `trim()`, aligning the build side with the lookup side and `getExactIdentityKey`.
- **new diagnostic**: `findDisabledAccountsWithFreshCredential` (read-only) surfaces disabled accounts holding an unexpired access token in both `codex-doctor` and `codex-health`, with six unit tests and the previously-requested `enabled: undefined` branch test added.

<h3>Confidence Score: 5/5</h3>

safe to merge — all three divergences are narrow, well-scoped fixes with mutation-tested coverage; no auth boundaries are widened and no writes are added.

the three changes each touch a single comparison or map-key site, guarded by the existing dedup/merge pipeline. sameEmailIdentity is only used at the email comparison site in isLegacyRefreshTokenDuplicate; accountId/organizationId comparisons via sameOptionalIdentity are untouched. sanitizeEmail(account.email) in buildIdentityIndexes is strictly more correct than bare trim() — the added @ validation is handled safely by the existing if(email) guard. findDisabledAccountsWithFreshCredential is purely read-only. all new paths have mutation-tested vitest coverage, the full suite passes, and tsc/eslint/build are clean.

no files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/storage/identity.ts | adds sameEmailIdentity (trim+lowercase) used in isLegacyRefreshTokenDuplicate, and lowercases the deduplicateAccountsByEmail map key; accountId/organizationId comparisons via sameOptionalIdentity are untouched |
| lib/auth/login-runner.ts | buildIdentityIndexes now keys byEmailNoOrg via sanitizeEmail (trim+lowercase+@ validation) instead of bare trim(), aligning the index-build side with the lookup side; sanitizeEmail returns undefined for undefined/empty/no-@ inputs so the if(email) guard still skips invalid entries correctly |
| lib/accounts/stale-state.ts | adds pure read-only findDisabledAccountsWithFreshCredential; enabled !== false guard correctly excludes enabled and undefined-enabled accounts |
| lib/tools/codex-doctor.ts | adds disabled-account-fresh-credential warning finding; storage null-guards the call consistently with the existing disabled-token-source-duplicate pattern |
| lib/tools/codex-health.ts | adds absorbedSlots to text output and disabledWithFreshCredentialSlots to JSON output; storage is guaranteed non-null here by the early-return guard at line 53 |
| test/stale-state.test.ts | six unit tests covering disabled+fresh, enabled, undefined-enabled, expired, token-less, and multi-slot ordering; the previously-flagged undefined branch is now covered |
| test/login-runner.test.ts | adds mixed-case merge-to-1 test and the previously-missing distinct-email control test; jwtWithEmail helper builds a valid base64url JWT with an email claim |
| test/storage.test.ts | adds org+token-dup case-differing email collapse test; correctly asserts single account, org identity preserved, and enabled !== false |

</details>

<sub>Reviews (2): Last reviewed commit: ["test: add enabled-undefined + distinct-e..."](https://github.com/ndycode/oc-codex-multi-auth/commit/4947d9e58f9c7ea409bf4ba2cbe0f8044838001f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38655326)</sub>

### CODE REVIEW

File: test/stale-state.test.ts [hunk 1]
```diff
@@ -168,3 +169,37 @@ describe("findStaleRecoverableAccounts", () => {
...
+
+	it("flags a disabled account that holds a fresh (unexpired) access token", () => {
+		const accounts = [{ enabled: false, accessToken: "tok", expiresAt: FUTURE }];
+		expect(findDisabledAccountsWithFreshCredential(accounts, NOW)).toEqual([0]);
+	});
+
+	it("ignores an enabled account (not the blind-spot case)", () => {
+		const accounts = [{ enabled: true, accessToken: "tok", expiresAt: FUTURE }];
+		expect(findDisabledAccountsWithFreshCredential(accounts, NOW)).toEqual([]);
+	});
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/stale-state.test.ts:185]: Added: "ignores an account with no enabled field (undefined is not disabled)" — asserts `findDisabledAccountsWithFreshCredential([{ accessToken, expiresAt: FUTURE }], NOW)` returns `[]`, putting teeth on the `enabled !== false` branch so a future guard refactor can't silently regress.

File: test/login-runner.test.ts [hunk 2]
```diff
@@ -49,6 +49,48 @@ describe("login-runner persistAccountPool", () => {
...
+					expires: Date.now() + 60_000,
+				},
+			],
+			false,
+		);
+		const second = await loadAccounts();
+		// Case-insensitive email index => merged into the existing entry, not appended.
+		expect(second?.accounts).toHaveLength(1);
+		expect(second?.accounts[0]?.refreshToken).toBe("refresh-new");
+	});
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/login-runner.test.ts:92]: Good catch — the control was only run ad-hoc, not committed. Added "keeps two genuinely distinct emails as separate accounts (control)": two sequential `persistAccountPool` calls with `user@example.com` + `other@example.com` (email-only, no accountId) assert `toHaveLength(2)`. Paired with the merge-to-1 test, this proves the email-dedup path is exercised and non-vacuous.
